src/plugins/telemetry/telemetry_impl.cpp

```diff
@@ -865,6 +882,65 @@ void TelemetryImpl::process_odometry(const mavlink_message_t& message)
     }
 }
 
+void TelemetryImpl::process_distance_sensor(const mavlink_message_t& message)
+{
+    Telemetry::DistanceSensor distance_sensor{};
+
+    distance_sensor.time_boot_ms = mavlink_msg_distance_sensor_get_time_boot_ms(&message);
+    distance_sensor.min_distance = mavlink_msg_distance_sensor_get_min_distance(&message);
+    distance_sensor.max_distance = mavlink_msg_distance_sensor_get_max_distance(&message);
+    distance_sensor.current_distance = mavlink_msg_distance_sensor_get_current_distance(&message);
+    distance_sensor.type = static_cast<Telemetry::DistanceSensor::SensorType>(mavlink_msg_distance_sensor_get_type(&message));
+    distance_sensor.id = mavlink_msg_distance_sensor_get_id(&message);
+    distance_sensor.covariance = mavlink_msg_distance_sensor_get_covariance(&message);
+    distance_sensor.horizontal_fov = mavlink_msg_distance_sensor_get_horizontal_fov(&message);
+    distance_sensor.vertical_fov = mavlink_msg_distance_sensor_get_vertical_fov(&message);
+
+    Telemetry::Quaternion quaternion{};
+    auto sensor_orientation = static_cast<MAV_SENSOR_ORIENTATION>(mavlink_msg_distance_sensor_get_orientation(&message));
+    switch (sensor_orientation) {
+        case MAV_SENSOR_ROTATION_NONE:
+            quaternion = {1, 0, 0, 0};
+            break;
+        case MAV_SENSOR_ROTATION_YAW_90:
+            quaternion = {std::sqrt(0.5f), 0, 0, std::sqrt(0.5f)};
+            break;
+        case MAV_SENSOR_ROTATION_YAW_270:
+            quaternion = {std::sqrt(0.5f), 0, 0, -1 * std::sqrt(0.5f)};
+            break;
+        case MAV_SENSOR_ROTATION_PITCH_180:
+            quaternion = {0, 0, 1, 0};
+            break;
+        case MAV_SENSOR_ROTATION_PITCH_90:
+            quaternion = {std::sqrt(0.5f), 0, std::sqrt(0.5f), 0};
+            break;
+        case MAV_SENSOR_ROTATION_PITCH_270:
+            quaternion = {std::sqrt(0.5f), 0, -1 * std::sqrt(0.5f), 0};
+            break;
+        case MAV_SENSOR_ROTATION_CUSTOM:
+        default:
+        {
+            std::array<float, 4> q{};
+            mavlink_msg_distance_sensor_get_quaternion(&message, q.data());
+            quaternion.w = q[0];
+            quaternion.x = q[1];
+            quaternion.y = q[2];
+            quaternion.z = q[3];
+            break;
+        }
+    }
```

Comment:
Should I add some other orientations?

Reply:
I don't know. Which orientations would PX4 ever send?

Reply:
If I'm reading this correctly there are two separate issues. One: vehicle orientation and as it relates to the quaternion and Two: returning distance sensor measurements?

 I'm looking more into Quaternions this evening. I was under the impression that the representation was a normalize vector representing a vehicle pointing and a rotation about that vector to add orientation. I was also under the impression that as long as I had the quaternion I could calculate vehicle attitude and sensor attitude in a fixed euclidian reference frame with a custom function or some off the shelf python package. 

The most useful thing to me would some sort of tree that contains the vehicle sensors and their values.

Reply:
> Which orientations would PX4 ever send?

Looks like this depends on real sensor orientation and we can't expect that it will be some constrained list. And in case of this orientation is not `CUSTOM` quaternion does not set.
The only way to implement safe message processing is check all orientations.

Reply:
> The most useful thing to me would some sort of tree that contains the vehicle sensors and their values.

@julianoes , @JonasVautherin what are you think about it? Can be implemented some around this in future?
This can be useful for me and I can help in implementation if it will start.

Reply:
I'm not sure I understand the "tree" idea there. You have multiple distance sensors on your drone, and for each of them you want to get the value they read. Correct? Why would that be a tree instead of a list?

The current proposition (subscribing to a stream of events that contains something like <sensor_id, sensor_value>) contains all the information I can imagine one may need. The only thing is that on the client side, you guys have to filter by id if you want only the events coming from one of the sensors. But that's trivial to do, right?

Reply:
@JonasVautherin I meant tree as in a tree of raw sensor readings. They could be any sensor type and distance sensor would be a subtype. Subtypes of distance sensor could be acoustic, Lidar, etc..

When you subscribe to a stream is that automatic or can MAVSDK return a list of what we can subscribe to and then decide? Say if we wanted to conserve bandwidth and only subscribe to streams we need.

Reply:
> When you subscribe to a stream is that automatic or can MAVSDK return a list of what we can subscribe to and then decide?

I would return the question to MAVLink: because MAVSDK uses MAVLink, it depends on how that works there. I am not sure if you can selectively register to the distance sensor you want using MAVLink. Can you?

Reply:
> I would return the question to MAVLink: because MAVSDK uses MAVLink, it depends on how that works there. I am not sure if you can selectively register to the distance sensor you want using MAVLink. Can you?

From a MAVLink perspective, in theory you can call [MAV_CMD_SET_MESSAGE_INTERVAL](https://mavlink.io/en/messages/common.html#MAV_CMD_SET_MESSAGE_INTERVAL) to control the rate of message emission. I'm not sure how widely this is supported in flight stacks, but on PX4 it is common to set the [mode](https://docs.px4.io/master/en/peripherals/mavlink_peripherals.html) to minimal and then specify the set of messages you want. 

FWIW I would think it reasonable for MAVSDK API to provide convenient filters on types/sensor ids you could register against. But I don't think it should be used for controlling sending rates as suggested here. That's more of an integration question.

Reply:
> FWIW I would think it reasonable for MAVSDK API to provide convenient filters on types/sensor ids you could register against.

I am not against the idea, but my concern is auto-generation of that kind of code. Because C++ is (still) not auto-generated, C++ implementers have more freedom now, but they will (hopefully soon) be constrained by the auto-generation. So when possible, I'm trying to make sure that it will be possible to auto-generate the APIs we add now. Once everything is auto-generated, it will be time to think about improving them.

Does that make sense?